packages/voting/src/denom.rs

```diff
@@ -0,0 +1,226 @@
+use cosmwasm_std::{Addr, Deps, StdError};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum DenomError {
+    #[error(transparent)]
+    Std(#[from] StdError),
+
+    #[error("invalid cw20 - did not respond to `TokenInfo` query: {err}")]
+    InvalidCw20 { err: StdError },
+
+    #[error("invalid native denom. length must be between in [3, 128], got ({len})")]
+    NativeDenomLength { len: usize },
+
+    #[error("expected alphabetic ascii character in native denomination")]
+    NonAlphabeticAscii,
+
+    #[error("invalid character ({c}) in native denom")]
+    InvalidCharacter { c: char },
+}
+
+/// A denom that has been checked to point to a valid asset. This enum
+/// should never be constructed literally and should always be built
+/// by calling `into_checked` on an `UncheckedDenom` instance.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum CheckedDenom {
+    /// A native (bank module) asset.
+    Native(String),
+    /// A cw20 asset.
+    Cw20(Addr),
+}
```

Comment:
This is great. We may as well break it out into another package?

Reply:
great idea! made `cw-denom`.